File: csharp/Pidigits.cs
using System.Numerics;
using System.Text;

public class Pidigits : Benchmark
{
    private int _nn;
    private StringBuilder _resultBuilder;

    public override uint Checksum => Helper.Checksum(_resultBuilder.ToString());

    public Pidigits()
    {
        _resultBuilder = new StringBuilder();
        _nn = (int)ConfigVal("amount");
    }

    public override void Run(long IterationId)
    {
        int i = 0;
        int k = 0;
        ulong ns = 0;
        BigInteger a = 0;
        BigInteger t;
        BigInteger u;
        int k1 = 1;
        BigInteger n = 1;
        BigInteger d = 1;

        while (true)
        {
            k += 1;
            t = n << 1;
            n *= k;
            k1 += 2;
            a = (a + t) * k1;
            d *= k1;

            if (a >= n)
            {
                var temp = n * 3 + a;
                var quotient = temp / d;
                t = quotient;
                u = temp % d;
                u += n;

                if (d > u)
                {
                    ns = ns * 10 + (ulong)t;
                    i += 1;

                    if (i % 10 == 0)
                    {
                        _resultBuilder.AppendFormat("{0:D10}\t:{1}\n", ns, i);
                        ns = 0;
                    }

                    if (i >= _nn)
                        break;

                    a = (a - (d * t)) * 10;
                    n *= 10;
                }
            }
        }

        if (ns != 0)
        {
            _resultBuilder.AppendFormat("{0:D10}\t:{1}\n", ns, i);
        }
    }
}

File: fsharp/Pidigits.fs
namespace Benchmarks

open System
open System.Numerics
open System.Text

type Pidigits() =
    inherit Benchmark()

    let mutable nn = 0
    let mutable resultBuilder = StringBuilder()

    override this.Run(_: int64) =
        let mutable i = 0
        let mutable k = 0
        let mutable ns = 0UL
        let mutable a = BigInteger.Zero
        let mutable k1 = 1
        let mutable n = BigInteger.One
        let mutable d = BigInteger.One
        let mutable finished = false

        while not finished do
            k <- k + 1
            let t = n <<< 1
            n <- n * BigInteger k
            k1 <- k1 + 2
            a <- (a + t) * BigInteger k1
            d <- d * BigInteger k1

            if a >= n then
                let temp = n * BigInteger 3 + a
                let quotient = temp / d
                let t2 = quotient
                let u = temp % d
                let u2 = u + n

                if d > u2 then
                    ns <- ns * 10UL + (uint64 t2)
                    i <- i + 1

                    if i % 10 = 0 then
                        resultBuilder.AppendFormat("{0:D10}\t:{1}\n", ns, i) |> ignore
                        ns <- 0UL

                    if i >= nn then
                        if ns <> 0UL then
                            resultBuilder.AppendFormat("{0:D10}\t:{1}\n", ns, i) |> ignore
                        finished <- true
                    else
                        a <- (a - (d * t2)) * BigInteger 10
                        n <- n * BigInteger 10

    override this.Checksum = 
        Helper.Checksum(resultBuilder.ToString())
    override this.Name = "CLBG::Pidigits"

    override this.Prepare() =
        nn <- Helper.Config_i64("CLBG::Pidigits", "amount") |> int
        resultBuilder <- StringBuilder()